### PR TITLE
docs(builtin): fix and annotate language blocks

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:639]:
```suggestion
			echo byteidx('a😊😊', 2)	    " returns 5
			echo byteidx('a😊😊', 2, 1)	" returns 1
			echo byteidx('a😊😊', 3, 1)	" returns 5
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:404]:
```suggestion
			blob2list(0z0102.0304)	" returns [1, 2, 3, 4]
			blob2list(0z)           " returns []
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:540]:
```suggestion
			bufname("#")		" alternate buffer name
			bufname(3)		" name of buffer 3
			bufname("%")		" name of current buffer
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:733]:
```suggestion
			char2nr(" ")		" returns 32
			char2nr("ABC")		" returns 65
			char2nr("á")		" returns 225
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:765]:
```suggestion
			charcol('.')		" returns 3
			col('.')		" returns 7
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:802]:
```suggestion
			echo charidx('áb́ć', 3)		" returns 1
			echo charidx('áb́ć', 6, 1)	" returns 4
			echo charidx('áb́ć', 16)		" returns -1
			echo charidx('a😊😊', 4, 0, 1)	" returns 2
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:882]:
```suggestion
			col(".")		" column of cursor
			col("$")		" length of cursor line plus one
			col("'t")		" column of mark t
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:1774]:
```suggestion
<			make /path/runtime/doc/builtin.o  >vim
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:1884]:
```suggestion
<			0  >vim
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:1886]:
```suggestion
<			1
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:2598]:
```suggestion
			getcharpos('.')		" returns [0, 5, 3, 0]
			getpos('.')		" returns [0, 5, 7, 0]
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:2801]:
```suggestion
			getcursorcharpos()	" returns [0, 3, 2, 0, 3]
			getcurpos()		" returns [0, 3, 4, 0, 3]
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:4544]:
```suggestion
			line(".")		" line number of the cursor
			line(".", winid)	" idem, in window "winid"
			line("'t")		" line number of mark t
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:4586]:
```suggestion
			list2blob([1, 2, 3, 4])	" returns 0z01020304
			list2blob([])		" returns 0z
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:4600]:
```suggestion
			list2str([32])		" returns " "
			list2str([65, 66, 67])	" returns "ABC"
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:4608]:
```suggestion
			list2str([97, 769])	" returns "á"
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:5550]:
```suggestion
			nr2char(64)		" returns "@"
			nr2char(32)		" returns " "
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:5552]:
```suggestion
			nr2char(300)		" returns I with bow character
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:8269]:
```suggestion
			str2list(" ")		" returns [32]
			str2list("ABC")		" returns [65, 66, 67]
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:8275]:
```suggestion
			str2list("á")		" returns [97, 769]
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:8402]:
```suggestion
		  :echo strftime("%c")		   " Sun Apr 27 11:49:23 1997
		  :echo strftime("%Y %b %d %X")	   " 1997 Apr 27 11:53:25
		  :echo strftime("%y%m%d %T")	   " 970427 11:53:55
		  :echo strftime("%H:%M")	   " 11:55
		  :echo strftime("%c", getftime("file.c"))
						   " Show mod time of file.c.
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:8435]:
```suggestion
		  :echo stridx("An Example", "Example")	     " 3
		  :echo stridx("Starting point", "Start")    " 0
		  :echo stridx("Starting point", "start")   " -1
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:8595]:
```suggestion
		    echo strutf16len('a')		" returns 1
		    echo strutf16len('©')		" returns 1
		    echo strutf16len('😊')		" returns 2
		    echo strutf16len('ą́')		" returns 1
		    echo strutf16len('ą́', v:true)	 " returns 3
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:9381]:
```suggestion
			echo utf16idx('a😊😊', 3)	" returns 2
			echo utf16idx('a😊😊', 7)	" returns 4
			echo utf16idx('a😊😊', 1, 0, 1)	" returns 2
```

File: runtime/doc/builtin.txt
Comment (clason) [inline runtime/doc/builtin.txt:9385]:
```suggestion
			echo utf16idx('aą́c', 6, 1)	" returns 4
			echo utf16idx('a😊😊', 9)	" returns -1
```

**Review by clason — COMMENTED** [review]
Sorry for commenting on the generated file (but I'd have messed up the whitespace anyway...)

Does it make sense to normalize the snippets to start with or (probably easier/better) without colon?

**Comment by clason** [comment]
And don't forget the static section "2. Matching a pattern in a String" (in `./scripts/gen_eval_files.lua`)!

**Comment by clason** [comment]
Random tangent (but don't want to forget): if we're touching `builtin.txt`, we can finally fix the invalid links `|v:_null_{blob,dict,list,string}|` that need excluding in `scripts/gen_help_html.lua`.

**Comment by lewis6991** [comment]
There should only be 2 ts errors by my estimation.

**Comment by clason** [comment]
As long as `scripts/gen_help_html.lua` doesn't complain, we're good!

**Review by clason — APPROVED** [review]

**Comment by clason** [comment]
oh boy, I should have tested this first -- `:h builtin` becomes unusable with all those injections 😞

(Looks nice, tho -- as a slideshow...)

**Comment by clason** [comment]
Regarding the parsing errors: The parser expects the closing `<` marker to be at column 0 here.

**Comment by lewis6991** [comment]
> oh boy, I should have tested this first -- `:h builtin` becomes unusable with all those injections 😞
> 
> (Looks nice, tho -- as a slideshow...)
> 
> Guess we have a strong incentive to revisit the injection performance improvements now 😬 

Oh no, I didn't notice. I hope it isn't too bad.

**Comment by clason** [comment]
It's horrible. As in, I need to `kill nvim` if I do `:h builtin` because the editor becomes unresponsive for minutes.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/24508 happens to remove many of the codeblocks in builtin.txt, so that should help performance.

**Comment by justinmk** [comment]
Should I wait on this before merging  https://github.com/neovim/neovim/pull/24508 ?

**Comment by lewis6991** [comment]
I'm undecided. The perf issues might mean we should sit on this for a bit.

Could you test this and see what it's like for you? I didn't notice any particular slowdown myself.

**Comment by clason** [comment]
> I'm undecided. The perf issues might mean we should sit on this for a bit.
> 
> Could you test this and see what it's like for you? I didn't notice any particular slowdown myself.

The perf issues I mention require the `comment` injections from nvim-treesitter; this means you'll have to have both `vim` and `comment` parsers installed using that plugin.

**Comment by lewis6991** [comment]
Ah that makes sense as I removed the comment parser.

Why are there even comment injections for vimdoc?

**Comment by clason** [comment]
Not for vimdoc, for vimscript (which is injected into vimdoc now).

File: runtime/queries/vimdoc/injections.scm
Comment (clason) [inline runtime/queries/vimdoc/injections.scm:4]: That will break the lua->vim injection in our other docs, which I would hate to lose?

Comment (lewis6991) [reply runtime/queries/vimdoc/injections.scm:4]: Are they really worth it?

The alternative (right now) would be to have no injections.

We can re-enable this when we add lazy injections.

Comment (clason) [reply runtime/queries/vimdoc/injections.scm:4]: Yes, they are really worth it. I'd rather drop the comment parser from nvim-treesitter.

Comment (lewis6991) [reply runtime/queries/vimdoc/injections.scm]: ok

**Comment by clason** [comment]
> Should I wait on this before merging #24508 ?

Depends on how hard it is to rebase this PR (i.e., how much manual work was necessary and has to be repeated). Otherwise there are independent benefits of #24508 (cleaner docs) and no drawbacks, so it would make sense to just go ahead with it already.

(Temporarily disabling the comment->vimscript injections would be the simplest workaround to unblock this.)

File: src/nvim/eval.lua
Comment (clason) [inline src/nvim/eval.lua:9331]:
```suggestion
      <Example bash command to list all Nvim servers: >bash
```

**Comment by lewis6991** [comment]
I say we just go with this. I'm petrified of having to deal with merge conflicts from this.

**Comment by clason** [comment]
Brace yourself then!

**Pull request merged**
